packages/@sanity/base/src/change-indicators/overlay/connectorPath.ts

```diff
@@ -0,0 +1,153 @@
+import {
+  ARROW_MARGIN_X,
+  ARROW_MARGIN_Y,
+  ARROW_SIZE,
+  ARROW_THRESHOLD,
+  CORNER_RADIUS
+} from '../constants'
+import {ConnectorLine} from './types'
+
+export function arrowPath(x: number, y: number, dir: number) {
+  return [
+    `M ${x - ARROW_SIZE} ${y - ARROW_SIZE * dir} `,
+    `L ${x} ${y}`,
+    `L ${x + ARROW_SIZE} ${y - ARROW_SIZE * dir}`
+  ].join('')
+}
+
+function moveTo(x: number, y: number) {
+  return `M${x} ${y}`
+}
+
+function lineTo(x: number, y: number) {
+  return `L${x} ${y}`
+}
+
+function join(strings: string[], delim = '') {
+  return strings.join(delim)
+}
+
+function quadCurve(x1: number, y1: number, x: number, y: number) {
+  return `Q${x1} ${y1} ${x} ${y}`
+}
+
+export function generateConnectorPath(line: ConnectorLine) {
+  const {from, to} = line
+  const {left: fromX, top: fromY} = from
+  const {left: toX, top: toY} = to
+
+  const cmds: string[] = []
+
+  // Calculate maximum corner radius
+  const r1 = Math.min(CORNER_RADIUS, Math.abs(fromY - toY) / 2)
+
+  // FROM
+  if (from.isAbove) {
+    cmds.push(
+      moveTo(fromX + ARROW_MARGIN_X, fromY - ARROW_THRESHOLD + ARROW_MARGIN_Y),
+      lineTo(fromX + ARROW_MARGIN_X, fromY - CORNER_RADIUS),
+      quadCurve(fromX + ARROW_MARGIN_X, fromY, fromX + ARROW_MARGIN_X + CORNER_RADIUS, fromY)
+    )
+  } else if (from.isBelow) {
+    cmds.push(
+      moveTo(fromX + ARROW_MARGIN_X, fromY + ARROW_THRESHOLD - ARROW_MARGIN_Y),
+      lineTo(fromX + ARROW_MARGIN_X, fromY + CORNER_RADIUS),
+      quadCurve(fromX + ARROW_MARGIN_X, fromY, fromX + ARROW_MARGIN_X + CORNER_RADIUS, fromY)
+    )
+  } else {
+    cmds.push(moveTo(fromX, fromY))
+  }
+
+  // TO
+  if (to.isAbove) {
+    if (fromY < to.bounds.top) {
+      cmds.push(
+        lineTo(to.bounds.left - 8 - r1, fromY),
```

Comment:
Something we could maybe do to make this piece of code even more easier to read is perhaps to divide up into "generate"-style functions so that we can have a specific name for each *segment* of the path. That way this could be something like:

```typescript
if (to.isAbove) {
  if (fromY < to.bounds.top) {
    lineSegments.push(
      generateCorner(from, to, ...)
    )
```

Basically "grouping" curves and lines into segments so we can reuse more and have more expressive names :)

Reply:
Yes, totally agree. Let's keep it as a future improvement?